core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,205 @@
+    }
+
+    // We explicitly state that while JSR-305 TYPE_USE annotations effectively should be applied to every type.
+    // They are not applicable for type parameter bounds because it would be a breaking change otherwise.
+    private fun Set<AnnotationQualifierApplicabilityType>.allIfTypeUse(): Set<AnnotationQualifierApplicabilityType> =
+        if (AnnotationQualifierApplicabilityType.TYPE_USE in this)
+            AnnotationQualifierApplicabilityType.values().toSet() - AnnotationQualifierApplicabilityType.TYPE_PARAMETER_BOUNDS + this
+        else
+            this
+
+    private fun resolveTypeQualifierDefaultAnnotation(annotation: Annotation): TypeQualifierWithApplicability<Annotation>? {
+        if (javaTypeEnhancementState.jsr305.isDisabled) return null
+        val typeQualifierDefault = annotation.findAnnotation(TYPE_QUALIFIER_DEFAULT_FQNAME) ?: return null
+        val typeQualifier = annotation.annotations.firstOrNull { resolveTypeQualifierAnnotation(it) != null } ?: return null
+        val applicability = typeQualifierDefault.enumArguments(onlyValue = true)
+            .mapNotNullTo(mutableSetOf()) { JAVA_APPLICABILITY_TYPES[it] }
+        return TypeQualifierWithApplicability(typeQualifier, applicability.allIfTypeUse())
+    }
+
+    fun isTypeUseAnnotation(annotation: Annotation): Boolean {
+        // Expect that Java's Target was mapped to Kotlin's Target.
+        val target = annotation.findAnnotation(StandardNames.FqNames.target) ?: return false
+        return target.enumArguments(onlyValue = false).any { it == KotlinTarget.TYPE.name }
+    }
+
+    private fun resolveJsr305AnnotationState(annotation: Annotation): ReportLevel {
+        resolveJsr305CustomState(annotation)?.let { return it }
+        return javaTypeEnhancementState.jsr305.globalLevel
+    }
+
+    private fun resolveJsr305CustomState(annotation: Annotation): ReportLevel? {
+        javaTypeEnhancementState.jsr305.userDefinedLevelForSpecificAnnotation[annotation.fqName]?.let { return it }
+        val enumValue = annotation.findAnnotation(MIGRATION_ANNOTATION_FQNAME)?.enumArguments(onlyValue = false)?.firstOrNull()
+            ?: return null
+        return javaTypeEnhancementState.jsr305.migrationLevel ?: when (enumValue) {
+            "STRICT" -> ReportLevel.STRICT
+            "WARN" -> ReportLevel.WARN
+            "IGNORE" -> ReportLevel.IGNORE
+            else -> null
+        }
+    }
+
+    private fun extractNullability(
+        annotation: Annotation, forceWarning: Annotation.() -> Boolean
+    ): NullabilityQualifierWithMigrationStatus? {
+        knownNullability(annotation, annotation.forceWarning())?.let { return it }
+
+        val typeQualifierAnnotation = resolveTypeQualifierAnnotation(annotation) ?: return null
+        val jsr305State = resolveJsr305AnnotationState(annotation)
+        if (jsr305State.isIgnore) return null
+        // TODO: the result of `forceWarning` will be overwritten - expected? Probably not.
```

Comment:
Yes, it looks strange. At least I'd replace `typeQualifierAnnotation.forceWarning()` with `jsr305State.isWarning`.